source/adios2/toolkit/transport/file/FileHTTP.cpp

```diff
@@ -0,0 +1,343 @@
+            for (; c < iovcnt; ++c)
+            {
+                Write(static_cast<const char *>(iov[c].iov_base),
+                      iov[c].iov_len);
+            }
+        }
+    };
+
+    WaitForOpen();
+    if (start != MaxSizeT)
+    {
+        errno = 0;
+        const auto newPosition = lseek(m_FileDescriptor, start, SEEK_SET);
+        m_Errno = errno;
+
+        if (static_cast<size_t>(newPosition) != start)
+        {
+            helper::Throw<std::ios_base::failure>(
+                "Toolkit", "transport::file::FilePOSIX", "WriteV",
+                "couldn't move to start position " + std::to_string(start) +
+                    " in file " + m_Name + " " + SysErrMsg());
+        }
+    }
+
+    int cntTotal = 0;
+    while (cntTotal < iovcnt)
+    {
+        int cnt = iovcnt - cntTotal;
+        if (cnt > 8)
+        {
+            cnt = 8;
+        }
+        lf_Write(iov + cntTotal, cnt);
+        cntTotal += cnt;
+    }
+}
+#endif
+
+void FileHTTP::Read(char *buffer, size_t size, size_t start)
+{
+    /* not using BUFSIZ, the server might use another value for that */
+    const size_t BUF_SIZE = 8192;
+    enum CONSTEXPR { MAX_REQUEST_LEN = 1024};
+    char request[MAX_REQUEST_LEN] = {'\0'};
+    int request_len =
+        snprintf(request, MAX_REQUEST_LEN, request_template, m_Name.c_str(),
+                 m_hostname.c_str(), start, start + size - 1);
+    if (request_len >= MAX_REQUEST_LEN) {
+        fprintf(stderr, "request length large: %d\n", request_len);
+        exit(EXIT_FAILURE);
+    }
+    m_socketFileDescriptor = socket(AF_INET, SOCK_STREAM, m_p_proto);
+    if (m_socketFileDescriptor == -1) {
+        perror("socket");
+        exit(EXIT_FAILURE);
+    }
+    /* Actually connect. */
+    if (connect(m_socketFileDescriptor, (struct sockaddr*)&sockaddr_in, sizeof(sockaddr_in)) == -1) {
+        perror("connect");
+        exit(EXIT_FAILURE);
+    }
+    /* Send HTTP request. */
+    int nbytes_total = 0;
+    while (nbytes_total < request_len) {
+        int nbytes_last = write(m_socketFileDescriptor, request + nbytes_total, request_len - nbytes_total);
+        if (nbytes_last == -1) {
+            perror("write");
+            exit(EXIT_FAILURE);
+        }
+        nbytes_total += nbytes_last;
+    }
+
+    /* Read the response. */
+    size_t bytes_recd = 0;
+    while (bytes_recd < size){
+        nbytes_total = read(m_socketFileDescriptor, buffer + bytes_recd, std::min(size - bytes_recd, BUF_SIZE));
+        if (nbytes_total == -1) {
+            perror("read");
+            exit(EXIT_FAILURE);
+        }
+        bytes_recd += nbytes_total;
+    }
+
+    close(m_socketFileDescriptor);
+    return;
+
+}
+
+size_t FileHTTP::GetSize()
+{
+    char request_template[] = "GET %s HTTP/1.1\r\nHost: %s\r\nContent-Length: bytes\r\n\r\n";
+    enum CONSTEXPR { MAX_REQUEST_LEN = 1024, BUF_SIZE = 128};
+    char request[MAX_REQUEST_LEN];
+    char buffer[BUF_SIZE] ={'\0'};
+    int request_len =
```

Comment:
Pretty sure that this code is wrong. snprintf() won't return more than MAX_REQUEST_LEN here.  (Or ever more than the limit it was given.). Instead, it'll return negative indicating an error.   Not sure why we'd call exit() on error anyway rather than throwing an exception.

Reply:
I made a small test program:
```#include <stdio.h>
int main()
{
        char buf[10];
        char template[] = "%s0123456789";
        int len = snprintf(buf, 10, template, "0000" );
        printf("%d\n", len);
        printf("%s\n", buf);
}
```
The output is 
```
14
000001234
```
It looks like sprint returns the number of characters that would have been written on the buffer, if buffer size had been sufficiently large.

Reply:
That behavior is more standard now, but hasn't always been and still isn't in some situations (VS has two versions of snprintf(), one of which keeps the old behavior (-1 return on hitting max) and one of which is new.  But this is all kind of beside the point.  Why would you use snprintf() in C++ anyway?  Why is the request_template a non-constant member of the class **_and_** a stack variable in a method?  Why on earth are the constants declared as enums?  Why is the code to do an http request essentially duplicated in the GetSize and Read methods?  Why doesn't the read() call in the while() loop have it's buffer and length values adjusted (like the write() does in the loop above)?  (Both read() and write() might result in partial completion requiring you do it in a while loop until it completes.  The write loop looks OK, but the read loop won't work if the read() doesn't complete in one read() (additional reads would overwrite the values read earlier rather than being added behind them).